docs/blockstates/extendedStates.md

```diff
@@ -0,0 +1,82 @@
+
+Declaring Unlisted Properties
+------------------------------------
+
+Unlisted properties are declared in `Block.createBlockState`, the same place as regular ("listed") properties. Instead of returning a `BlockStateContainer`, one must return an `ExtendedBlockState`. Forge provides a builder `BlockStateContainer.Builder`, which will automatically handle returning an `ExtendedBlockState` for you.
+
+Example:
+```Java
+@Override
+public BlockStateContainer createBlockState() {
+	return new BlockStateContainer.Builder(this).add(LISTED_PROP).add(UNLISTED_PROP).build();
+}
+```
+
+Filling Extended States
+----------------------------
+
+Before an `IBlockState` is used for rendering purposes, it will always have `Block.getExtendedState` called on it first. In this method, you will give all your unlisted properties values. Assuming you registered at least one unlisted property in the previous section, the `IBlockState` parameter can be safely casted to `IExtendedBlockState`, which has a `withProperty` method for unlisted properties analogous to its listed property cousin. Here, you can query whatever you want from the World, the Tile Entity, etc. (with appropriate safety checks, of course) and insert it into the extended blockstate.
+
+!!! Warning
+	It is highly recommended that your unlisted property values be immutable. Baked model implementations will use the extended state and unlisted values on multiple threads, so any value must be used in a threadsafe manner. The easiest way is to simply make that information an immutable snapshot.
+	Because of this, `Block.getExtendedState` is the last opportunity you have on the main thread to pass information to the rendering system. Anything you might possibly want to know in your custom `IBakedModel`, you should be passing an immutable snapshot of through `Block.getExtendedState`.
+
+Example:
+```Java
+@Override
+public IExtendedBlockState getExtendedState(IBlockState state, IBlockAccess world, BlockPos pos) {
+	IExtendedBlockState ext = (IExtendedBlockState) state;
+	TileEntity te = world.getTileEntity(pos);
+	if (te instanceof MyTE) {
+		ext = ext.withProperty(UNLISTED_PROP, ((MyTE) te).getSomeImmutableData());
+	}
+	return ext;
+}
+```
+
+Using Extended States
+--------------------------
+
+In a custom `IBakedModel`, the `IBlockState` parameter passed to you will be exactly the object you returned in `Block.getExtendedState`, and you can pull data out of it and use it to affect your rendering as you wish.
+
+Here is a basic example of using an unlisted property to determine which model to render, assuming `UNLISTED_PROP` is an `IUnlistedProperty<String>`:
+```Java
+// in custom IBakedModel
+private final Map<String, IBakedModel> submodels = new HashMap<>(); // populated in a custom manner out of the scope of this article
+
+@Override
+public List<BakedQuad> getQuads(IBlockState state, EnumFacing facing, long rand) {
+	IExtendedBlockState ex = (IExtendedBlockState) state;
```

Comment:
It's probably good practice to nullcheck the state, as it's a `@Nullable` parameter.